metagraph/src/graph/alignment/dbg_aligner.cpp

```diff
@@ -38,110 +39,462 @@ ::ISeedAndExtendAligner(const DeBruijnGraph &graph, const DBGAlignerConfig &conf
         };
 
         auto get_min_path_score = [&](const Alignment &seed) {
-            return aggregator.get_min_path_score(seed);
+            return std::max(config_.min_path_score,
+                            aggregator.get_min_path_score(seed));
         };
 
-        std::string_view this_query = paths.get_query(is_reverse_complement);
+        std::string_view this_query = paths[i].get_query(is_reverse_complement);
         assert(this_query == query);
 
-        std::vector<node_index> nodes = map_to_nodes_sequentially(graph_, query);
-
-        auto seeder = build_seeder(this_query, is_reverse_complement, nodes);
-        auto extender = build_extender(this_query, aggregator);
-
-        size_t num_explored_nodes = 0;
+        auto extender = build_extender(this_query, aggregator, config_);
 
 #if ! _PROTEIN_GRAPH
-        if (graph_.get_mode() == DeBruijnGraph::CANONICAL
-                || config_.forward_and_reverse_complement) {
-            std::vector<node_index> nodes_rc(nodes);
-            std::string dummy(query);
-            reverse_complement_seq_path(graph_, dummy, nodes_rc);
-            assert(dummy == paths.get_query(!is_reverse_complement));
-            assert(nodes_rc.size() == nodes.size());
-
-            std::string_view reverse = paths.get_query(!is_reverse_complement);
-
-            auto seeder_rc = build_seeder(reverse, !is_reverse_complement, nodes_rc);
-            auto extender_rc = build_extender(reverse, aggregator);
+        if (seeder_rc) {
+            std::string_view reverse = paths[i].get_query(!is_reverse_complement);
+            auto extender_rc = build_extender(reverse, aggregator, config_);
 
-            align_both_directions(this_query, reverse, *seeder, *seeder_rc,
-                                  *extender, *extender_rc,
-                                  add_alignment, get_min_path_score);
+            auto [seeds, extensions, explored_nodes] =
+                align_both_directions(this_query, reverse, *seeder, *seeder_rc,
+                                      *extender, *extender_rc,
+                                      add_alignment, get_min_path_score);
 
-            num_explored_nodes += extender_rc->num_explored_nodes();
+            num_seeds += seeds;
+            num_extensions += extensions;
+            num_explored_nodes += explored_nodes;
 
         } else {
-            align_core(this_query, *seeder, *extender, add_alignment, get_min_path_score);
+            align_core(*seeder, *extender, add_alignment, get_min_path_score, false);
         }
 #else
-        align_core(this_query, *seeder, *extender, add_alignment, get_min_path_score);
+        align_core(*seeder, *extender, add_alignment, get_min_path_score, false);
 #endif
 
         num_explored_nodes += extender->num_explored_nodes();
+        num_extensions += extender->num_extensions();
+        size_t num_labels = std::max(size_t { 1 }, aggregator.num_labels() - 1);
```

Comment:
Why does this have such a weird definition?

Reply:
fixed